pkg/controller/external.go

```diff
@@ -162,25 +162,28 @@ func (e *external) Observe(ctx context.Context, mg xpresource.Managed) (managed.
 		return managed.ExternalObservation{}, errors.Wrap(err, "cannot get connection details")
 	}
 
+	lateInitedParams, err := tr.LateInitialize(res.State.GetAttributes())
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, "cannot late initialize parameters")
+	}
+	lateInitialized := annotationsUpdated || lateInitedParams
+	resourceReady := tr.GetCondition(xpv1.TypeReady).Equal(xpv1.Available())
 	// We try to mark the resource ready before the spec late initialization
 	// and a call for "Plan" because those operations are costly, i.e. late-init
 	// causes a spec update which defers status update to the next reconcile and
-	// "Plan" takes a few seconds.
-	if !tr.GetCondition(xpv1.TypeReady).Equal(xpv1.Available()) {
+	// "Plan" takes a few seconds. We assume that a successful state refresh
+	// with state info available for the remote infrastructure in the output
+	// terraform.tfstate indicates readiness.
+	if !resourceReady || lateInitialized {
 		tr.SetConditions(xpv1.Available())
 		return managed.ExternalObservation{
 			ResourceExists:          true,
 			ResourceUpToDate:        true,
 			ConnectionDetails:       conn,
-			ResourceLateInitialized: annotationsUpdated,
+			ResourceLateInitialized: lateInitialized && resourceReady,
```

Comment:
Even if late-initialization is done, we should issue status update, i.e. `ResourceLateInitialized: false`, because readiness is more important than updating the spec. The only exception is the critical annotations we can't loose like external name.

Reply:
@muvaf thanks for the clarification for updating the critical annotations. I have opened this PR for further discussion:
https://github.com/crossplane/terrajet/pull/270